File: exercise/java/day19/src/test/java/CalculatorTests.java
import static org.assertj.core.api.Assertions.*;
import static travel.SantaTravelCalculator.calculateTotalDistanceRecursively;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.ValueSource;

class CalculatorTests {

    @ParameterizedTest
    @CsvSource({
            "1, 1",
            "2, 3",
            "5, 31",
            "10, 1023",
            "20, 1048575",
            "30, 1073741823"
    })
    void shouldCalculateTheDistanceFor(int numberOfReindeers, int expectedDistance) {
        assertThat(calculateTotalDistanceRecursively(numberOfReindeers)).isEqualTo(expectedDistance);
    }

    @ParameterizedTest
    @ValueSource(ints = {32, 50})
    void failForNumbersGreaterThan32(int numberOfReindeers) {
        assertThatThrownBy(() -> calculateTotalDistanceRecursively(numberOfReindeers))
                .isInstanceOf(ArithmeticException.class)
                .hasMessage("Integer overflow");
    }
}

File: exercise/kotlin/day19/src/test/kotlin/CalculatorTests.kt
package travel.tests

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FunSpec
import io.kotest.datatest.withData
import io.kotest.matchers.shouldBe
import travel.SantaTravelCalculator

class CalculatorTests : FunSpec({
    context("should calculate the distance correctly") {
        withData(
            1 to 1,
            2 to 3,
            5 to 31,
            10 to 1023,
            20 to 1048575,
            30 to 1073741823
        ) { (numberOfReindeers, expectedDistance) ->
            SantaTravelCalculator.calculateTotalDistanceRecursively(numberOfReindeers) shouldBe expectedDistance
        }
    }

    context("should fail for numbers greater than 32") {
        withData(32, 50) { numberOfReindeers ->
            shouldThrow<ArithmeticException> {
                SantaTravelCalculator.calculateTotalDistanceRecursively(numberOfReindeers)
            }.message shouldBe "Integer overflow"
        }
    }
})